Content.Server/Chat/Systems/ChatSystem.cs

```diff
@@ -483,41 +487,58 @@ private void OnGameChange(GameRunLevelChangedEvent ev)
 
             if (MessageRangeCheck(session, data, range) != MessageRangeCheckResult.Full)
                 continue; // Won't get logged to chat, and ghosts are too far away to see the pop-up, so we just won't send it to them.
 
+            var canUnderstand = _language.CanUnderstand(listener, language);
+            var finalMessage = canUnderstand ? message : languageObfuscatedMessage;
+
             if (data.Range <= WhisperClearRange)
-                _chatManager.ChatMessageToOne(ChatChannel.Whisper, message, wrappedMessage, source, false, session.Channel);
-            //If listener is too far, they only hear fragments of the message
-            //Collisiongroup.Opaque is not ideal for this use. Preferably, there should be a check specifically with "Can Ent1 see Ent2" in mind
-            else if (_interactionSystem.InRangeUnobstructed(source, listener, WhisperMuffledRange, Shared.Physics.CollisionGroup.Opaque)) //Shared.Physics.CollisionGroup.Opaque
-                _chatManager.ChatMessageToOne(ChatChannel.Whisper, obfuscatedMessage, wrappedobfuscatedMessage, source, false, session.Channel);
-            //If listener is too far and has no line of sight, they can't identify the whisperer's identity
+            {
+                var wrappedResult = Loc.GetString("chat-manager-entity-whisper-wrap-message",
+                    ("entityName", name),
+                    ("message", FormattedMessage.EscapeText(finalMessage)));
+
+                // If the listener is in the clear range, do not perform further obfuscations
+                _chatManager.ChatMessageToOne(ChatChannel.Whisper, finalMessage, wrappedResult, source, false, session.Channel);
+            }
+            else if (_interactionSystem.InRangeUnobstructed(source, listener, WhisperMuffledRange, Shared.Physics.CollisionGroup.Opaque))
+            {
+
+                // If the listener is too far, they only hear fragments of the message
+                // Collisiongroup.Opaque is not ideal for this use. Preferably, there should be a check specifically with "Can Ent1 see Ent2" in mind
+                var result = ObfuscateMessageReadability(finalMessage);
+                var wrappedResult = Loc.GetString("chat-manager-entity-whisper-wrap-message",
+                    ("entityName", nameIdentity), ("message", FormattedMessage.EscapeText(result)));
+
+                _chatManager.ChatMessageToOne(ChatChannel.Whisper, result, wrappedResult, source, false, session.Channel);
+            }
             else
-                _chatManager.ChatMessageToOne(ChatChannel.Whisper, obfuscatedMessage, wrappedUnknownMessage, source, false, session.Channel);
+            {
+                //If listener is too far and has no line of sight, they can't identify the whisperer's identity
+                var result = ObfuscateMessageReadability(finalMessage);
+                var wrappedResult = Loc.GetString("chat-manager-entity-whisper-unknown-wrap-message",
```

Comment:
Please try not to include 'magic strings' in your code like this. Any strings being accessed in code should generally be defined as consts at the top of the script to make debugging for others not a nightmare.

Reply:
It's not even our code, just a part of reorganized upstream code- :sob:

How would you recommend to do it?

Reply:
Eh, honestly leave it in that case. Refactors can happen later. The only point here then would be to try and reorganize *less* upstream code without refactoring it.

For reference, the correct way to hardcode a value is to not hardcode a value, but to leave it as a DataField in a Component somewhere. The second best way to hardcode a value is to do it as a 'private const string <name> = <string>;` at the very top of the file, to make these values clear and collected.